### ISSUE TITLE
Tensorboard logging by epoch instead of by step

### ISSUE THREAD

**Issue by mpaepper** [PyTorchLightning/pytorch-lightning#2110]
Short question concerning the tensorboard logging:

I am using it like this:

```python
    def training_epoch_end(self, outputs):
        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
        tensorboard_logs = {'train/loss': avg_loss}
        for name in self.metrics:
            tensorboard_logs['train/{}'.format(name)] = torch.stack([x['metr'][name] for x in outputs]).mean()

        return {'loss': avg_loss, 'log': tensorboard_logs}
```

It works very well, but in the plots (the x-axis) is the step, so each batch is a step. Is it possible to have the x-axis be the epoch as I want to plot the metrics only per epoch?

**Comment by awaelchli** [comment]
use a different logger. 
TensorBoard can't do this stuff.

**Comment by mpaepper** [comment]
Thank you for your answer, but I don't think that TensorBoard can't do it.

I am currently migrating from Torchbearer and there I am also using Tensorboard and it shows the epoch on the x-axis instead of a step, so it is possible.

**Comment by mpaepper** [comment]
To clarify: I am referring just to changing the x-axis numbers from steps to epochs, so I don't have to calculate in which epochs something happened.
![Auswahl_178](https://user-images.githubusercontent.com/4135790/84013653-7fe3e480-a979-11ea-9723-853c9bd42cb4.png)

**Comment by awaelchli** [comment]
Yes, because they are using the global step as the epoch, right?

**Comment by awaelchli** [comment]
I guess you could try this: 
1. return log: {'global_step': self.current_epoch} and nothing else in your training_step
2. return  your metrics in training_epoch_end

**Comment by mpaepper** [comment]
Thank you, that was a good hint. I debugged it now.

It's possible to pass in the 'step' which will be used as the current_epoch like this:

```python
    def training_epoch_end(self, outputs):
        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
        tensorboard_logs = {'train/loss': avg_loss}
        for name in self.metrics:
            tensorboard_logs['train/{}'.format(name)] = torch.stack([x['metr'][name] for x in outputs]).mean()
        tensorboard_logs['step'] = self.current_epoch

        return {'loss': avg_loss, 'log': tensorboard_logs}
```

**Comment by awaelchli** [comment]
ok, great. did you also try to log additionally in the training_step method? Because I think that would not work. Just saying in case someone else finds this issue later on.

**Comment by mpaepper** [comment]
No, I am only logging in the end of epoch method and didn't try otherwise.

**Comment by adeboissiere** [comment]
Hi, 

It is possible to track both the steps and epochs using tensorboard. Here is an example. It is quite straightforwd. 

```python
    def training_step(self, batch, batch_idx):
        batch, y = batch
        y_hat = self(batch)

        labels_hat = torch.argmax(y_hat, dim=1)
        n_correct_pred = torch.sum(y == labels_hat).item()

        loss = F.cross_entropy(y_hat, y.long())
        tensorboard_logs = {'train_acc_step': n_correct_pred, 'train_loss_step': loss}

        return {'loss': loss, "n_correct_pred": n_correct_pred, "n_pred": len(y), 'log': tensorboard_logs}

    def training_epoch_end(self, outputs):
        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()

        train_acc = sum([x['n_correct_pred'] for x in outputs]) / sum(x['n_pred'] for x in outputs)
        tensorboard_logs = {'train_acc': train_acc, 'train_loss': avg_loss, 'step': self.current_epoch}

        return {'loss': avg_loss, 'log': tensorboard_logs}

    def validation_step(self, batch, batch_idx):
        batch, y = batch
        y_hat = self(batch)

        loss = F.cross_entropy(y_hat, y.long())
        labels_hat = torch.argmax(y_hat, dim=1)
        n_correct_pred = torch.sum(y == labels_hat).item()

        return {'val_loss': loss, "n_correct_pred": n_correct_pred, "n_pred": len(y)}

    def validation_epoch_end(self, outputs):
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()

        val_acc = sum([x['n_correct_pred'] for x in outputs]) / sum(x['n_pred'] for x in outputs)
        tensorboard_logs = {'val_loss': avg_loss, 'val_acc': val_acc, 'step': self.current_epoch}

        return {'log': tensorboard_logs}

    def test_step(self, batch, batch_idx):
        batch, y = batch
        y_hat = self(batch)

        loss = F.cross_entropy(y_hat, y.long())
        labels_hat = torch.argmax(y_hat, dim=1)
        n_correct_pred = torch.sum(y == labels_hat).item()

        return {'test_loss': loss, "n_correct_pred": n_correct_pred, "n_pred": len(y)}

    def test_epoch_end(self, outputs):
        avg_loss = torch.stack([x['test_loss'] for x in outputs]).mean()
        test_acc = sum([x['n_correct_pred'] for x in outputs]) / sum(x['n_pred'] for x in outputs)
        tensorboard_logs = {'test_loss': avg_loss, 'test_acc': test_acc, 'step': self.current_epoch}

        return {'log': tensorboard_logs}

```

Here is what it looks like.

![image](https://user-images.githubusercontent.com/11660643/84429215-c8a7d180-abf5-11ea-9f8b-1940031666dc.png)

Cheers!

**Comment by awaelchli** [comment]
I did not know! Thanks @adeboissiere. So in summary we can provide the step with 
`{"step": self.current_epoch}` or whatever we want.

**Comment by dschaehi** [comment]
Does the solution still work with the latest version 0.10.0?

**Comment by george-gca** [comment]
I solved this using this code. Note that, in my case, I may run my code with only one or multiple validation datasets:

```python
def validation_epoch_end(self, outputs):
    if not self.trainer.running_sanity_check:
        for dataset_result in outputs:
            # in case of using only one validation dataset
            if type(dataset_result) is dict:
                for key, val in dataset_result.items():
                    if is_tensor(val):
                        dataset_result[key] = val.cpu().detach()

                self.logger.agg_and_log_metrics(
                    dataset_result, step=self.current_epoch)
                    
            # in case of using multiple validation datasets
            else:  # if type(dataset_result) is list:
                for metrics in dataset_result:
                    for key, val in metrics.items():
                        if is_tensor(val):
                            metrics[key] = val.cpu().detach()

                    self.logger.agg_and_log_metrics(
                        metrics, step=self.current_epoch)
```

**Comment by eagomez2** [comment]
As of today returning a dict with the 'log' key is deprecated, is there any other solution to preserve the right x-axis numbering?

**Comment by george-gca** [comment]
Have you tried getting the [logger](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#logger) and manually calling [log_metrics](https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.loggers.TensorBoardLogger.html#pytorch_lightning.loggers.TensorBoardLogger.log_metrics) with the right step?

**Comment by matejciglenecki** [comment]
I create a simple callback `OverrideEpochStepCallback` which overrides the `step` to the current epoch index. `step` is overridden on every _epoch-ending_ procedure:

```py
from pytorch_lightning.callbacks import Callback
import pytorch_lightning as pl

class OverrideEpochStepCallback(Callback):
    def __init__(self) -> None:
        super().__init__()

    def on_training_epoch_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule):
        self._log_step_as_current_epoch(trainer, pl_module)

    def on_test_epoch_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule):
        self._log_step_as_current_epoch(trainer, pl_module)

    def on_validation_epoch_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule):
        self._log_step_as_current_epoch(trainer, pl_module)

    def _log_step_as_current_epoch(self, trainer: pl.Trainer, pl_module: pl.LightningModule):
        pl_module.log("step", trainer.current_epoch)

trainer = pl.Trainer(callbacks=[OverrideEpochStepCallback()])
```

This way, you don't need to implement `training_epoch_end`. `training_step` can log metrics every epoch and the `OverrideEpochStepCallback` takes care of the step number:
```py
def training_step(self, batch, batch_idx):
        x, y = batch
        y_pred = self(x)

        loss = F.cross_entropy(y_pred, y)
        acc = multi_acc(y_pred, y)

        data_dict = {
            "loss": loss,  # the 'loss' key needs to be present
            "train/loss": loss,
            "train/acc": acc,
        }
        log_dict = data_dict.copy() # we don't want to log `loss` metric
        log_dict.pop("loss", None)
        self.log_dict(log_dict, on_step=True, on_epoch=True, logger=True, prog_bar=True) # step and epoch logs both work!
        return data_dict
```

**Comment by Petros626** [comment]
> Thank you, that was a good hint. I debugged it now.
> 
> It's possible to pass in the 'step' which will be used as the current_epoch like this:
> 
> ```python
>     def training_epoch_end(self, outputs):
>         avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
>         tensorboard_logs = {'train/loss': avg_loss}
>         for name in self.metrics:
>             tensorboard_logs['train/{}'.format(name)] = torch.stack([x['metr'][name] for x in outputs]).mean()
>         tensorboard_logs['step'] = self.current_epoch
> 
>         return {'loss': avg_loss, 'log': tensorboard_logs}
> ```

Where to insert this piece of code?

**Comment by GoldenEXPn** [comment]
> For anyone who would like to re-use the `OverrideEpochStepCallback` mentioned above, **replace `on_training_epoch_end` with `on_train_epoch_end`**.

Bro thanks you saved my life
